contents/stacks_and_queues/code/c++/queue.cpp

```diff
@@ -0,0 +1,100 @@
+#include<iostream>
+using namespace std;
+
+namespace my {
+    /**
+     * parameterised type implementation using linked list
+     * [value][next] -> [value][next] -> ... -> [value][next] -> [value][next]
+     *  (front Node)   (intermediat Nodes)    (rear Node)      (dummy Node)
+     */
+    template<typename T>
+    struct Node {
+        /**
+        * next: will store right Node address
+        */
+        T value;
+        Node<T>* next;
+        Node(const T& V) : value(V), next(nullptr) { }
+    };
+
+    template<typename T>
+    class queue {
+    private:
+        /**
+         * _front :  points to left most node
+         * count: keeps track of current number of elements present in queue excluding dummy Node
+         * rear:  points to most recent Node added into the queue, which is just left size of dummy Node
+         */
+        Node<T>* _front;
+        Node<T>* rear;
+        size_t count;
+    public:
+        queue() : _front(nullptr), rear(nullptr), count(0ULL) {
+            _front = rear = new Node<T>(0); // creating a dummy Node
+        }
+
+        void push(const T& element) {
+            Node<T>* new_node = new Node<T>(element);  // create New Node
+            if (count > 0) {
+                new_node->next = rear->next; // make new Node point to dummy Node
+                rear->next = new_node;       // make rear Node point to new Node
+                rear = new_node;             // make rear Node's pointer to point to new Node
+            } else {
+                new_node->next = rear;
+                rear = _front = new_node;
+            }
+            count = count + 1;
+        }
+
+        void dequeue() {
+            if (count > 0) {
+                Node<T>* buffer = _front;
+                _front = _front->next;
+                count = count - 1;
+                delete buffer;
+            }
+            rear = (count != 0) ? rear : _front;
+        }
+
+        T& front() const { return _front->value; }
```

Comment:
I would also add the version with `T const& front() const`
This allows to use `front()` in cases where a `const`-reference is needed.

```suggestion
        T& front() const { return _front->value; }
        T const& front() const { return _front->value; }
```